alicloud/data_source_alicloud_api_gateway_groups.go

```diff
@@ -0,0 +1,152 @@
+package alicloud
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/cloudapi"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceAlicloudApiGatewayGroups() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAlicloudApigatewayGroupsRead,
+
+		Schema: map[string]*schema.Schema{
+			"name_regex": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateNameRegex,
+			},
+			"output_file": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			// Computed values
+			"groups": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"region_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"sub_domain": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"created_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"modified_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"traffic_limit": {
+							Type:     schema.TypeInt,
+							Computed: true,
+						},
+						"billing_status": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"illegal_status": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+func dataSourceAlicloudApigatewayGroupsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AliyunClient).cloudapiconn
+
+	args := cloudapi.CreateDescribeApiGroupsRequest()
+	args.RegionId = string(getRegion(d, meta))
+	args.PageSize = requests.NewInteger(PageSizeLarge)
+	args.PageNumber = requests.NewInteger(1)
+
+	var allGroups []cloudapi.ApiGroupAttribute
+
+	for {
+		resp, err := conn.DescribeApiGroups(args)
+		if err != nil {
+			return err
+		}
+
+		if resp == nil {
+			break
+		}
+
+		allGroups = resp.ApiGroupAttributes.ApiGroupAttribute
```

Comment:
I think line 99 should be `allGroups = append(allGroups, resp.ApiGroupAttributes.ApiGroupAttribute...)`